sync/optimistic.md

```diff
@@ -0,0 +1,290 @@
+# Optimistic Sync
+
+## Introduction
+
+In order to provide a syncing execution engine with a partial view of the head
+of the chain, it may be desirable for a consensus engine to import beacon
+blocks without verifying the execution payloads. This partial sync is called an
+*optimistic sync*.
+
+## Constants
+
+|Name|Value|Unit
+|---|---|---|
+|`SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY`| `128` | slots
+
+*Note: the `SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY` must be user-configurable. See
+[Fork Choice Poisoning](#fork-choice-poisoning).*
+
+## Helpers
+
+Let `head_block: BeaconBlock` be the result of calling of the fork choice
+algorithm at the time of block production.
+
+Let `justified_block: BeaconBlock` be the latest current justified ancestor
+ancestor of the `head_block`.
+
+Let `optimistic_roots: Set[Root]` be the set of `hash_tree_root(block)` for all
+optimistically imported blocks which have yet to receive an `INVALID` or
+`VALID` designation from an execution engine.
+
+```python
+def is_optimistic(block: BeaconBlock) -> bool:
+    hash_tree_root(block) in optimistic_roots
+```
+
+```python
+def latest_valid_ancestor(block: BeaconBlock) -> BeaconBlock:
+    while True:
+        if not is_optimistic(block) or block.parent_root == Root():
+	        return block
+        block = get_block(block.parent_root)
+```
+
+```python
+def is_execution_block(block: BeaconBlock) -> BeaconBlock:
+	block.body.execution_payload != ExecutionPayload()
+```
+
+```python
+def should_optimistically_import_block(current_slot: Slot, block: BeaconBlock) -> bool:
+	block.slot + SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY <= current_slot
+```
+
+Let only a node which returns `is_optimistic(head) == True` be an *optimistic
+node*. Let only a validator on an optimistic node be an *optimistic validator*.
+
+When this specification only defines behaviour for an optimistic
+node/validator, but *not* for the non-optimistic case, assume default
+behaviours without regard for optimistic sync.
+
+## Mechanisms
+
+## When to optimistically import blocks
+
+A block MUST NOT be optimistically imported, unless either of the following
```

Comment:
This can probably be reworded to more straightforward language:
```suggestion
A block can be optimistically imported ONLY when either of the following conditions are met
```

Reply:
Ah yeah, I was trying to work within the [RFC keywords](https://datatracker.ietf.org/doc/html/rfc2119). I'll check it out and see if I can make it simpler :pray: